### DISCUSSION TITLE
min_cycle_duration VS control_output

### DISCUSSION THREAD

**Discussion by sciorty** [ScratMan/HASmartThermostat#183] [Q&A]
How does the **min_cycle_duration** behaves against the control_output?
If i want to set it to, let's say, 30 minutes, if the control output generates a period of 1 minute, does this mean that the heater will not be switched on until it will generate at least 30 minutes, or does it mean that the heater will be turned on for 30 minutes no matter what the control output says (as long as it's >0)?

Also, i'm trying to understand how to calibrate the parameters but the control output has a different behaviour from what i expected:

Configuration:
```
 - platform: smart_thermostat
    name: Termostato Cameretta
    unique_id: termostato_cameretta
    heater: switch.riscaldamento_cameretta
    target_sensor: sensor.thp_cameretta_temperatura
    min_temp: 10
    max_temp: 26
    target_temp_step: 0.1 
    ac_mode: False
    away_temp: 18
    cold_tolerance: 0
    precision: 0.1
    keep_alive:
      seconds: 60    

    kp: 5
    ki: 0.1
    kd: 500
    
    debug: true
    sampling_period: 00:10:00
    pwm: 1:00:00
```

Debug:
```
hvac_modes:
  - heat
  - 'off'
min_temp: 10
max_temp: 26
target_temp_step: 0.1
preset_modes:
  - none
  - away
current_temperature: 20
temperature: 21.5
hvac_action: idle
preset_mode: none
away_temp: 18
eco_temp: null
boost_temp: null
comfort_temp: null
home_temp: null
sleep_temp: null
activity_temp: null
control_output: 26.6
kp: 5
ki: 0.01
kd: 500
ke: 0
pid_mode: auto
pid_i: 19.3
pid_p: 7.4
pid_d: -0.1
pid_e: 0
pid_dt: 600.0040857791901
friendly_name: Termostato Cameretta
supported_features: 17
```

If my total output is 26%, with a pwm of 1h, why is the heater off? Shouldn't it turn the heater on for 26 minutes?

**Comment by wiktor-z** [comment]
From my observations, it works differently: if you set min_cycle_duration to 30 minutes and pwm set to 60 minutes, then each heating activation will last AT LEAST 30 minutes... for example: control_output = 10 -> it will activate heating for 6 minutes (because 10% of 60 minutes is 6 minutes), but due to min_cycle_duration = 30 minutes, the heating will be activated for 30 minutes. Only in the case when control_output has a value greater than 50 in this example, the heating will be activated for more than half an hour.

So, what is the purpose of min_cycle_duration? Some heating devices (such as gas boilers) don't like frequent turning on and off, so, for example, min_cycle_duration=5 might be a good idea, but you must verify this with the instructions for your heating device.

And one more thing, control_output = 26 at pwm=60 is not 26 minutes, but 26% of 60 minutes, which is 15 minutes and 36 seconds. If you prefer to think of the control output in terms of time, you can consider setting the PWM to 01:40:00, which is equivalent to 100 minutes.

**Reply by sciorty** [reply]
Thanks, so that’s the second guess i had. Sorry for the minutes of the pwm, i got them wrong.

I’m now really satisfied with the thermostat, as you can see it just have errors of +- 0.1 °C, but i still don’t see the relation between the control output and the heater state. 
As you can see in my graph the control output is always greater than zero, with a pwm of one hour, so being the control output usually around 5 or 10, those “needles” of 3-6 minutes of heating are right. But i really can’t understan when they do “start”? I mean, when is the output applied? When it is generated? At the end of the pwm?.. or what?
![IMG_0168](https://github.com/ScratMan/HASmartThermostat/assets/6460914/4b934b22-38c4-4b77-9ad9-b57c35b42083)
![IMG_0169](https://github.com/ScratMan/HASmartThermostat/assets/6460914/51dfc245-530f-4060-bacc-58fcf1f332d8)

**Comment by wiktor-z** [comment]
The last graph shows that your pid_i is always greater than zero and continuously contributes to the control output. The integral part is the component that takes into account the system's past state, summing up the past errors. You can observe the behavior of your setup when you reset an integral part by calling the 'smart_thermostat: Clear Integral' service. HOWEVER, this action may destabilize your setup. I'd be pretty amazed if I managed to hit a difference of 0.1 degrees Celsius in my setup. At that point, I'd likely just leave the configuration alone ;)

**Reply by sciorty** [reply]
Yeah i understand how the control output is computed, what i don’t understand is WHEN it is applied. If it were applied every time it is computed, i should have the heater always on (as it is always >0), don’t you agree? 
So i’m missing something to fully understand the behaviour. Maybe the remaining part of the duty cycle is supposed to be heater off? Example: control output 10%, pwm of 1 hour -> heater on for 6 mins and then off for 54 mins no matter of the control output value on the remaining 54 minutes? It would make sense but as you can see the numbers do not match..

**Reply by ScratMan** [reply]
If output is lower than minimum duration, the pwm duration is stretched to respect the output ratio.
Example: pwm is 60mn, min duration is 10mn. If control output is 10% (6mn) the pwm will be stretched to 100mn and switch on the heating for 10mn and then switch off for 90mn.

**Reply by sciorty** [reply]
Thanks @ScratMan, this replies to both my questions. Anyway, in the example of a 60 minutes PWM, you meant ON for 10 minutes and then off for 50 minutes, didn't you?

So is it correct to say that, in the same example of a PWM of 60 minutes, that the heating cycle will be applied only every 60 minutes? Using your example, the next ON/OFF period will be determined at the end of the 50 OFF minutes (so after 60 minutes from the ON period), looking at the control output value in that exact time and "ignoring" the other values?

**Reply by ScratMan** [reply]
No, the output (and pwm) is evaluated continuously when an input sensor is updated, and depending on the keep alive and sampling period settings if any.

**Reply by sciorty** [reply]
Then i don’t understant:
keeping the same example, supposing we start at a time = 00:00:00, with pwm of 01:00:00.
The control output is computed and equal to 10%, so it will turn the heater on for 6 minutes and then off for 54. Right?
What’s next? I should have now the entire pwm cycle of 1h “covered and programmed”.
Supposing that at time t=00:00:04 a sensor data arrives increasing the control output to 20%, what happens? Are this 12 resulting minutes added to the 6 minutes computed before? What if the control output is instead down to 5%, will this make the heater turn off before the 6 minutes programmed? 

I’m really confused more than before..
